### PR TITLE
Add list itemLayoutAnimation documentation page and example

### PR DESCRIPTION
**Pull request by MatiPl01**
## Summary

This PR adds example usage of `itemLayoutAnimation` prop to the example app and docs page with example recording and details.

## Example image of the added docs page

Screenshot 2024-07-22 at 15 46 23

## Related context

Related issue: #6278

Support for `multi-column` lists seems to be impossible to implement. react-native adds additional wrapper component for each row and re-renders list items in different rows when new items are added or items are removed from the list. Because the parent of list items changes and the layout animation cannot be applied to the wrapper that is added to list rows, layout animations won't work for lists with multiple columns.

At least, I didn't come up with any valid solution.

PR that adds support for `FlatList` items animations: #2674
What react-native does for mutli-column lists: https://github.com/facebook/react-native/blob/2098806c2207f376027184329a7285913ef8d090/packages/react-native/Libraries/Lists/FlatList.js#L643

### CODE REVIEW

File: packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+```
+
+</div>
+
+</Row>
+
+## Remarks
+
+- `itemLayoutAnimation` works only with a single-solumn `Animated.FlatList`,
+- it does not work when `numColumns` is greater than 1,
```
Comment (tjzel) [inline packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx:44]:
```suggestion
- `itemLayoutAnimation` works only with a single-column `Animated.FlatList`, `numColumns` property cannot be grater than 1.
```

File: packages/react-native-reanimated/src/component/FlatList.tsx [hunk 2]
```diff
@@ -45,6 +45,7 @@ interface ReanimatedFlatListPropsWithLayout<T>
   extends AnimatedProps<FlatListProps<T>> {
   /**
    * Lets you pass layout animation directly to the FlatList item.
+   * (Works only with a single-column Animated.FlatList - when numColumns is not greater than 1)
```
Comment (tjzel) [inline packages/react-native-reanimated/src/component/FlatList.tsx:48]:
```suggestion
   * Works only with a single-column `Animated.FlatList`. `numColumns` property cannot be greater than 1.
```

File: packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx [hunk 3]
```diff
@@ -0,0 +1,54 @@
+---
+sidebar_position: 6
+title: List Layout Animations
+sidebar_label: List Layout Animations
+---
+
+`itemLayoutAnimation` lets you define the animation that will be used when list items layout changes. You can use one of the predefined [layout transitions](/docs/layout-animations/layout-transitions#predefined-transitions) like `LinearTransition` or create your own [custom layout transition](/docs/layout-animations/custom-animations#custom-layout-transition).
```
Comment (tjzel) [inline packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx:7]:
```suggestion
`itemLayoutAnimation` lets you define a [layout transition](/docs/layout-animations/layout-transitions) that's applied when list items layout changes. You can use one of the [predefined transitions](/docs/layout-animations/layout-transitions#predefined-transitions) like `LinearTransition` or create [your own transition](/docs/layout-animations/custom-animations#custom-layout-transition).
```

File: apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        return aNum - bNum;
+      });
+      return newItems;
+    });
+  }, []);
+
+  const transition = LinearTransition;
+
+  return (
+    // Skip initial entering animation
```
Comment (tjzel) [inline apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx:75]: A line below tells exactly the same.
```suggestion
    // Skip initial entering animation
```

File: apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    <Pressable onPress={() => onPress(id)} style={styles.listItem}>
+      <Text style={styles.itemText}>{text}</Text>
+    </Pressable>
+  );
+});
+
+export default function ListItemLayoutAnimation() {
+  const [items, setItems] = useState(ITEMS);
+
+  const handlePress = useCallback((id: string) => {
```
Comment (tjzel) [inline apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx:35]:
```suggestion
  const removeItem = useCallback((id: string) => {
```

File: apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    setItems((prevItems) => prevItems.filter((item) => item !== id));
+  }, []);
+
+  const renderItem = useCallback<ListRenderItem<string>>(
+    ({ item }) => <ListItem id={item} text={item} onPress={handlePress} />,
+    [handlePress]
+  );
+
+  const findItemName = useCallback(() => {
```
Comment (tjzel) [inline apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx:44]:
```suggestion
  const getNewItemName = useCallback(() => {
```

File: apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+      return newItems;
+    });
+  }, []);
+
+  const resetOrder = useCallback(() => {
+    setItems((prevItems) => {
+      const newItems = [...prevItems];
+      newItems.sort((a, b) => {
+        const aNum = parseInt(a.match(/\d+$/)![0], 10);
+        const bNum = parseInt(b.match(/\d+$/)![0], 10);
```
Comment (tjzel) [inline apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx:65]: A bit ugly but ok.
```suggestion
        const aNum = parseInt(a.match(/\d+$/)![0]);
        const bNum = parseInt(b.match(/\d+$/)![0]);
```

File: packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx
Comment (tjzel) [inline packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx:1]: Can you change `itemLayoutAnimation` on the fly? i.e.

`itemLayoutAnimation = { condition ? first : second }`

Please include the answer in the page.

**Review by tjzel — CHANGES_REQUESTED** [review]

File: apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+      const newItems = [...prevItems];
+      newItems.sort(() => Math.random() - 0.5);
+      return newItems;
+    });
+  }, []);
+
+  const resetOrder = useCallback(() => {
+    setItems((prevItems) => {
+      const newItems = [...prevItems];
+      newItems.sort((a, b) => {
```
Comment (tjzel) [inline apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx:63]:
```suggestion
      newItems.sort((left, right) => {
```

File: apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx (hunk 7 above)
Comment (MatiPl01) [reply apps/common-app/src/examples/LayoutAnimations/ListItemLayoutAnimation.tsx:65]: I know but eslint is very sad 😢 when this param is missing 

Screenshot 2024-07-22 at 15 05 23

File: packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx
Comment (MatiPl01) [reply packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx:1]: Ok, I added this to the example in the app and will add info to docs. I think that I don't have to update the recording, though. Just mentioning this possibility in docs will be enough.

https://github.com/user-attachments/assets/1302fadd-4dd5-4640-a993-fae0f77963b6

Comment (MatiPl01) [reply packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx:1]: I added changes to the example app and code example to the **Remarks** section. I don't think I need to update the recording, information in the **Remarks** section should be enough.

https://github.com/user-attachments/assets/dfe6e148-a2d6-4545-8637-84431e75af7e

File: packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+</div>
+
+</Row>
+
+## Remarks
+
+- `itemLayoutAnimation` works only with a single-column `Animated.FlatList`, `numColumns` property cannot be grater than 1.
+- you can change the `itemLayoutAnimation` on the fly or disable it by setting it to `undefined`.
```
Comment (tjzel) [inline packages/docs-reanimated/docs/layout-animations/list-layout-animations.mdx:44]:
```suggestion
- You can change the `itemLayoutAnimation` on the fly or disable it by setting it to `undefined`.
```

**Review by tjzel — APPROVED** [review]

**Pull request merged**